servers/rendering/renderer_rd/renderer_scene_render_forward.cpp

```diff
@@ -1837,113 +1948,158 @@ void RendererSceneRenderForward::_render_scene(RID p_render_buffer, const Transf
 		RD::get_singleton()->texture_resolve_multisample(render_buffer->color_msaa, render_buffer->color);
 	}
-}
 
-void RendererSceneRenderForward::_render_shadow(RID p_framebuffer, const PagedArray<GeometryInstance *> &p_instances, const CameraMatrix &p_projection, const Transform &p_transform, float p_zfar, float p_bias, float p_normal_bias, bool p_use_dp, bool p_use_dp_flip, bool p_use_pancake, const Plane &p_camera_plane, float p_lod_distance_multiplier, float p_screen_lod_threshold, const Rect2i &p_rect, bool p_flip_y, bool p_clear_region, bool p_begin, bool p_end) {
-	RENDER_TIMESTAMP("Setup Rendering Shadow");
+	RD::get_singleton()->draw_command_end_label();
+}
 
+void RendererSceneRenderForward::_render_shadow_begin() {
+	scene_state.shadow_passes.clear();
+	RD::get_singleton()->draw_command_begin_label("Shadow Setup");
 	_update_render_base_uniform_set();
 
-	render_pass++;
+	render_list[RENDER_LIST_SECONDARY].clear();
+}
+void RendererSceneRenderForward::_render_shadow(RID p_framebuffer, const PagedArray<GeometryInstance *> &p_instances, const CameraMatrix &p_projection, const Transform &p_transform, float p_zfar, float p_bias, float p_normal_bias, bool p_use_dp, bool p_use_dp_flip, bool p_use_pancake, const Plane &p_camera_plane, float p_lod_distance_multiplier, float p_screen_lod_threshold, const Rect2i &p_rect, bool p_flip_y, bool p_clear_region, bool p_begin, bool p_end) {
+	uint32_t shadow_pass_index = scene_state.shadow_passes.size();
+
+	SceneState::ShadowPass shadow_pass;
 
 	scene_state.ubo.dual_paraboloid_side = p_use_dp_flip ? -1 : 1;
 
-	_setup_environment(RID(), RID(), p_projection, p_transform, RID(), true, Vector2(1, 1), 1, 32, RID(), !p_flip_y, Color(), 0, p_zfar, false, p_use_pancake);
+	_setup_environment(RID(), RID(), p_projection, p_transform, RID(), true, Vector2(1, 1), 1, 32, RID(), !p_flip_y, Color(), 0, p_zfar, false, p_use_pancake, shadow_pass_index);
 
 	if (get_debug_draw_mode() == RS::VIEWPORT_DEBUG_DRAW_DISABLE_LOD) {
 		p_screen_lod_threshold = 0.0;
 	}
 
 	PassMode pass_mode = p_use_dp ? PASS_MODE_SHADOW_DP : PASS_MODE_SHADOW;
 
-	_fill_render_list(p_instances, pass_mode, p_projection, p_transform);
-
-	RID rp_uniform_set = _setup_render_pass_uniform_set(RID(), RID(), RID(), RID(), RID(), PagedArray<RID>(), PagedArray<RID>());
-
-	RENDER_TIMESTAMP("Render Shadow");
-
-	render_list.sort_by_key(false);
+	uint32_t render_list_from = render_list[RENDER_LIST_SECONDARY].elements.size();
+	_fill_render_list(RENDER_LIST_SECONDARY, p_instances, pass_mode, p_projection, p_transform, false, false, p_camera_plane, p_lod_distance_multiplier, p_screen_lod_threshold, true);
+	uint32_t render_list_size = render_list[RENDER_LIST_SECONDARY].elements.size() - render_list_from;
+	render_list[RENDER_LIST_SECONDARY].sort_by_key_range(render_list_from, render_list_size);
+	_fill_instance_data(RENDER_LIST_SECONDARY, render_list_from, render_list_size, false);
```

Comment:
While testing, I got a crash in `RendererSceneRenderForward::_fill_instance_data()` because it was being called from here with `render_list_size = 0`.

My understanding is that `_fill_render_list()` did not add any elements for some reason. I got the crash by changing the scale of a MeshInstance in the inspector, but I could only reproduce this once. 

@reduz Do you think we should just guard against this possibility? Or we should look into why `_fill_render_list()` didn't add any elements to the render list?

Reply:
It should not be a problem, I tested with zero elements and seems to work ok, if you can figure out how to reproduce it again that would be great, else I guess we will eventually rely on user reports to track this down.

Reply:
To be more precise, it crashed at `renderer_scene_render_forward.cpp` line 1300.

`total_elements` was 10, the size of `instance_data` was also 10, and `p_offset` was 10 as well. This made me think this line:
`uint32_t element_total = p_max_elements > 0 ? p_max_elements : rl->elements.size();`
may be incorrect, because in this case there were 0 elements in total. 

Maybe the check needs to be `p_max_elements >= 0` and the default argument value should be -1?

Reply:
Look like it was the correct fix, then? :)